sdk/go/pulumi/stack_reference.go

```diff
@@ -13,14 +15,26 @@ type StackReference struct {
 	Name StringOutput `pulumi:"name"`
 	// Outputs resolves with exports from the named stack
 	Outputs MapOutput `pulumi:"outputs"`
+
+	// Engine context in which resolve resource references in outputs.
+	context *Context
 }
 
 // GetOutput returns a stack output keyed by the given name as an AnyOutput
 func (s *StackReference) GetOutput(name StringInput) AnyOutput {
-	return All(name, s.Outputs).
-		ApplyT(func(args []interface{}) interface{} {
-			n, outs := args[0].(string), args[1].(map[string]interface{})
-			return outs[n]
+	return All(name, s.rawOutputs).
+		ApplyT(func(args []interface{}) (AnyOutput, error) {
+			n, stack := args[0].(string), args[1].(resource.PropertyMap)
+			if !stack["outputs"].IsObject() {
+				return Any(nil), fmt.Errorf("failed to convert %T to object", stack)
+			}
+			outs := stack["outputs"].ObjectValue()
+			ret, secret, _ := unmarshalPropertyValue(s.context, outs[resource.PropertyKey(n)])
+
+			if secret {
+				ret = ToSecret(ret)
+			}
+			return Any(ret), nil
```

Comment:
This is all conflict with https://github.com/pulumi/pulumi/pull/9855. 😄

Reply:
Yours looked good, I tested the nonexistant key behavior, merged & then included tests in this PR :)